### PR TITLE
Added config entity for GraphQL query maps.

### PR DESCRIPTION
**Pull request by paranojik**
Modified the route enhanced to accept an additional parameter for query map version id and for loading the map from config instead of file.

### CODE REVIEW

File: src/Form/GraphQLQueryMapForm.php [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+   * {@inheritdoc}
+   */
+  public function form(array $form, FormStateInterface $form_state) {
+    $form = parent::form($form, $form_state);
+
+    $query_map = $this->entity;
+
+    $form['#title'] = $this->t('Edit %title GraphQL Query Map', ['%title' => $query_map->id()]);
+
+    $form['version_id'] = [
```
Comment (fubhy) [inline src/Form/GraphQLQueryMapForm.php]: Let's generate this instead of providing a form. No?

File: src/Form/GraphQLQueryMapForm.php [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    $form['query_map'] = [
+      '#type' => 'fieldset',
+      '#title' => $this->t('GraphQL Queries'),
+      '#prefix' => '<div id="names-fieldset-wrapper">',
+      '#suffix' => '</div>',
+      '#tree' => TRUE,
+    ];
+
+    $count = $form_state->get('query_map_count');
+    if (empty($count)) {
```
Comment (fubhy) [inline src/Form/GraphQLQueryMapForm.php]: Let's make this a .json file upload.

File: src/Form/GraphQLQueryMapForm.php [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    ];
+    $form['actions']['submit'] = [
+      '#type' => 'submit',
+      '#value' => $this->t('Submit'),
+    ];
+
+    return $form;
+  }
+
+  public function addOne(array &$form, FormStateInterface $form_state) {
```
Comment (fubhy) [inline src/Form/GraphQLQueryMapForm.php]: This is nice but realistically we would have an upload of a .json file and that would then be parsed and the queries from that json file woudl be added one by one.

File: src/Form/GraphQLQueryMapImportForm.php [hunk 4]
```diff
@@ -0,0 +1,77 @@
+<?php
+
+namespace Drupal\graphql\Form;
```
Comment (fubhy) [inline src/Form/GraphQLQueryMapImportForm.php:3]: Oh, now I see.. You built both! Nice. I don't think we need the other (manual) form for now. Can you split up this PR for just the upload form for now and submit the manual form separately? I need to think about that more.

File: src/QueryNotFoundException.php [hunk 5]
```diff
@@ -0,0 +1,10 @@
+<?php
+/**
+ * @file
+ * Contains \Drupal\graphql\QueryNotFoundException.
+ */
+
+namespace Drupal\graphql;
+
+
```
Comment (fubhy) [inline src/QueryNotFoundException.php]: Unnecessary line break.

File: src/QueryNotFoundException.php [hunk 6]
```diff
@@ -0,0 +1,10 @@
+<?php
+/**
+ * @file
```
Comment (fubhy) [inline src/QueryNotFoundException.php]: I don't want @file headers. They are redundant :).

Comment (fubhy) [reply src/QueryNotFoundException.php]: Actually, I am not sure if we even need a dedicated exception type. What do you think?

File: src/Routing/QueryRouteEnhancer.php [hunk 7]
```diff
@@ -100,13 +118,14 @@ protected function enhanceSingle($defaults, $values) {
       'query' => NULL,
       'variables' => NULL,
       'id' => NULL,
+      'versionId' => NULL,
```
Comment (fubhy) [inline src/Routing/QueryRouteEnhancer.php]: It's not a version Id. It's a hash. Let's just call it "version".

**Review by fubhy — COMMENTED** [review]
Very good job.  However, can you refactor the whole thing and replace all the snake_case variables to camelCase? I am using camelCase exclusively across the rest of the project. Thanks <3

File: src/Form/GraphQLQueryMapForm.php (hunk 1 above)
Comment (paranojik) [reply src/Form/GraphQLQueryMapForm.php]: Yeah... I was thinking about that... but what if you want to make some small adjustments without modifying the version id? Or does that not make sense?

File: src/Form/GraphQLQueryMapForm.php [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    // During the initial form build, add this form object to the form state and
+    // allow for initial preparation before form building and processing.
+    if (!$form_state->has('entity_form_initialized')) {
+      $this->init($form_state);
+    }
+
+    // Ensure that edit forms have the correct cacheability metadata so they can
+    // be cached.
+    if (!$this->entity->isNew()) {
+      \Drupal::service('renderer')->addCacheableDependency($form, $this->entity);
```
Comment (fubhy) [inline src/Form/GraphQLQueryMapForm.php]: Why are we using \Drupal here?

File: src/Form/GraphQLQueryMapImportForm.php [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  public function validateForm(array &$form, FormStateInterface $form_state) {
+    $files = file_save_upload('query_map_json', [
+      'file_validate_extensions' => ['json'], // Validate extensions.
+    ]);
+
+    if (!isset($files[0])) {
+      $form_state->setError($form['query_map_json'], $this->t('No file was uploaded.'));
```
Comment (fubhy) [inline src/Form/GraphQLQueryMapImportForm.php]: $formState

File: src/Form/GraphQLQueryMapImportForm.php [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+        '%id' => $graphqlQueryMap->id(),
+      ]));
+    }
+    else {
+      drupal_set_message($this->t('The GraphQL Query Map version %id was not saved.', [
+        '%id' => $graphqlQueryMap->id(),
+      ]));
+    }
+
+    $form_state->setRedirect('entity.graphql_query_map.collection');
```
Comment (fubhy) [inline src/Form/GraphQLQueryMapImportForm.php]: $formState

File: src/Routing/QueryRouteEnhancer.php [hunk 11]
```diff
@@ -53,7 +71,7 @@ protected function getValuesFromRequest(Request $request) {
 
     // Only keep numerical keys or 'query', 'variables', 'id'.
```
Comment (fubhy) [inline src/Routing/QueryRouteEnhancer.php]: Comment needs to be adjusted (now also 'version')

File: src/Routing/QueryRouteEnhancer.php [hunk 12]
```diff
@@ -98,22 +116,23 @@ protected function enhanceBatch($defaults, $values) {
...
-      'query' => '',
-      'variables' => [],
-      'id' => '',
-    ], array_flip(['query', 'variables', 'id']));
-
-    $query = $values['query'];
-    $id = $values['id'];
-    $variables = $values['variables'];
-    if (empty($query) && empty($id)) {
+      'query' => NULL,
```
Comment (fubhy) [inline src/Routing/QueryRouteEnhancer.php]: should be empty string by default.

File: src/Routing/QueryRouteEnhancer.php [hunk 13]
```diff
@@ -98,22 +116,23 @@ protected function enhanceBatch($defaults, $values) {
...
+      'variables' => NULL,
```
Comment (fubhy) [inline src/Routing/QueryRouteEnhancer.php]: should be empty array by default

File: src/Routing/QueryRouteEnhancer.php [hunk 14]
```diff
@@ -98,22 +116,23 @@ protected function enhanceBatch($defaults, $values) {
...
+      'id' => NULL,
+      'version' => NULL,
+    ], array_flip(['query', 'variables', 'id', 'version']));
+
+    $variables = $this->getVariables($values['variables']);
+    if (empty( $values['query']) && empty($values['id'])) {
```
Comment (fubhy) [inline src/Routing/QueryRouteEnhancer.php]: There is an extra whitespace before $values['query']

File: src/Routing/QueryRouteEnhancer.php [hunk 15]
```diff
@@ -98,22 +116,23 @@ protected function enhanceBatch($defaults, $values) {
...
+      return FALSE;
+    }
+
+    if (empty($values['query']) && empty($values['id']) && empty($values['version'])) {
       return FALSE;
     }
```
Comment (fubhy) [inline src/Routing/QueryRouteEnhancer.php]: I believe you didn't resolve the conflict in this method properly. I changed the code here a bit and you brought back some of the things I removed. Can you check the latest code in your merge target?

File: src/Routing/QueryRouteEnhancer.php [hunk 16]
```diff
@@ -98,22 +116,23 @@ protected function enhanceBatch($defaults, $values) {
...
+      'id' => NULL,
+      'version' => NULL,
+    ], array_flip(['query', 'variables', 'id', 'version']));
+
+    $variables = $this->getVariables($values['variables']);
+    if (empty( $values['query']) && empty($values['id'])) {
+      return FALSE;
+    }
+
+    if (empty($values['query']) && empty($values['id']) && empty($values['version'])) {
```
Comment (fubhy) [inline src/Routing/QueryRouteEnhancer.php]: This condition is incorrect. We either need a 'query' OR we need ('id' AND 'version'). Let's change it to this:

so: empty($values['query']) && (empty($values['id']) && empty($values['version']))

File: src/Routing/QueryRouteEnhancer.php [hunk 17]
```diff
@@ -122,4 +141,32 @@ protected function enhanceSingle($defaults, $values) {
...
-}
\ No newline at end of file
+
+  /**
+   * @param $query
+   * @param $id
+   * @param $version
+   * @return null
+   */
+  protected function getQuery($query, $id, $version) {
```
Comment (fubhy) [inline src/Routing/QueryRouteEnhancer.php:148]: if we have a $query, just return it. If we don't have a $query, load by $id and $version.

```
if (!empty($query)) {
  return $query;
}

$queryMap = $this->entityTypeManager->getStorage('graphql_query_map')->load($version);
return !empty($queryMap[$id]) ? $queryMap[$id] : NULL;
```

File: src/Routing/QueryRouteEnhancer.php [hunk 18]
```diff
@@ -122,4 +141,32 @@ protected function enhanceSingle($defaults, $values) {
...
+    }
+
+    return $query;
+  }
+
+  /**
+   * @param $variables
+   * @return array|mixed
+   */
+  protected function getVariables($variables) {
```
Comment (fubhy) [inline src/Routing/QueryRouteEnhancer.php]: This method was removed by me. Please check your conflict resolution.

**Review by fubhy — CHANGES_REQUESTED** [review]
We are getting there. Looking great.

**Pull request merged**